Support NTP server details in base, EOS

bewing opened the pull request:
Based on models from openconfig-system, return optional additional data regarding NTP servers for the first supported driver (EOS).

bewing committed:
Support non-total model testing in helper

In support of adding details to NTP server dictionaries,
support detection of non-total TypedDicts and change the logic
to only error if there are EXTRA keys instead of missing keys.

bewing committed:
Update NTP data return definition

Support returning details regarding individual NTP servers based
on openconfig models (/system/ntp/servers/server)

bewing committed:
Add support for NTP server details to EOS

bewing commented on napalm/base/models.py:
```diff
@@ -224,15 +224,31 @@
     "NTPPeerDict",
     {
-        # will populate it in the future wit potential keys
+        "address": str,
```
Adding the same keys here was a quick and dirty fix to avoid refactoring nxos's mypy typing for same.  I may revisit.

https://github.com/napalm-automation/napalm/actions/runs/12915579137/job/36017755840#step:7:14

ktbyers commented on napalm/base/test/helpers.py:
```diff
@@ -4,23 +4,26 @@
+    if model.__total__:
+        same_keys = set(annotations.keys()) == set(data.keys())
+    else:
+        same_keys = set(data.keys()) <= set(annotations.keys())
```
We generally haven't allowed there to be a subset of keys in data. Is there a way in testing that we can specifically allow this in this case.

The equivalent of:

```python
def test_model(model, data, allow_subset=False):
```

But where we pass in `allow_subset=True`, not sure how we would actualize that in the testing code.

ktbyers commented on napalm/base/test/helpers.py:
```diff
@@ -4,23 +4,26 @@
     correct_class = True
-    for key, instance_class in model.items():
-        correct_class = isinstance(data[key], instance_class) and correct_class
+    for key in data.keys():
```
Again we are switching over from the `annotations` to the `data` which is dangerous. Can we call out the `allow_subset=True` to specifically allow this in this case?

ktbyers reviewed (commented):
I only reviewed part of the testing here...but it is a start.

bewing committed:
Add kwarg for model subset testing

Since we only have a limited set of models that we are testing
subsets of keys for, we should explicitly only allow subsets when
testing those getters.

bewing committed:
Refactor nxos ntp for mypy typing

Making NTPPeerDict and NTPServerDict non-identical caused
issues with typing since we wre unable to assert the type of the
return dictionary easily.  This refactor allows us to explicitly
state the return type for each function.

bewing committed:
remove unused EOS NTP peer textfsm

ktbyers committed:
Merge branch 'develop' into ntp_details

ktbyers commented on napalm/base/models.py:
```diff
@@ -223,16 +223,22 @@
+        "prefer": bool,
+        "network_instance": str,
+        "source_address": str,
+        "key_id": int,
```
So these are all optional key-values which may be returned?

bewing replied:
That is correct.    If we get to the point where all drivers return it in all situations, we could make it non-optional.

ktbyers reviewed (approved):
LGTM
